### PR TITLE
Canbus mcp2515

### PR DESCRIPTION
**Pull request by mvturnho**
## Description:
Implementation to interconnect esphome devices using canbus.

This is the the canbus_mcp2515 component implementation .
You use it with the canbus_base platform component.
So to start testing this you also need the "canbus_base" platform component implementation .

you may test this code by checking out the full canbus that includes the canbus_base and canbus_mcp2515 implementation.
All code changes are first tested in canbus and then commited in canbus_base or canbus_mcp2515

canbus_base
canbus_mcp2515
canbus = canbus_base + canbus_mcp2515

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by mvturnho** [comment]
@OttoWinter If I can be of any additional service to get this commited please let me know.

File: esphome/components/mcp2515/mcp2515.cpp [hunk 1]
```diff
@@ -0,0 +1,641 @@
...
+  this->disable();
+
+  return i;
+}
+
+canbus::Error MCP2515::set_mode_(const CANCTRL_REQOP_MODE mode) {
+  modify_register_(MCP_CANCTRL, CANCTRL_REQOP, mode);
+
+  unsigned long endTime = millis() + 10;
+  bool modeMatch = false;
```
Comment (OttoWinter) [inline esphome/components/mcp2515/mcp2515.cpp]:
```suggestion
  bool mode_match = false;
```

See contributing guide and travis checks.

**Review by OttoWinter — COMMENTED** [review]
Hi,

The code itself looks ok, but the lint checks do not pass (they need to, otherwise I cannot merge). I think the provided scripts (namely quicklint) would fix most of these automatically. Please run them over the codebase (you can do so with docker too, see contributing guide)

**Comment by mvturnho** [comment]
@OttoWinter  as for the last error; the mcp2515 can not do without the canbus component. 
I can not solve thi in the split PR's for canbus and mcp2515

**Comment by mvturnho** [comment]
@OttoWinter fixed the virtual functions. But one issue remains since the canbus and mcp2515 are separate pullrequests.

```
45.45s$ pylint esphome
************* Module esphome.components.mcp2515.canbus
esphome/components/mcp2515/canbus.py:3:0: E0611: No name 'canbus' in module 'esphome.components' (no-name-in-module)
esphome/components/mcp2515/canbus.py:5:0: E0611: No name 'canbus' in module 'esphome.components' (no-name-in-module)
esphome/components/mcp2515/canbus.py:5:0: E0401: Unable to import 'esphome.components.canbus' (import-error)
```

**Pull request closed without merging**